numba/parfors/parfor.py

```diff
@@ -4064,28 +4078,232 @@ def remove_duplicate_definitions(blocks, nameset):
+                if fbs_res is not None:
+                    ind_seq, _ = fbs_res
+                    # If any indices are derived from an index is used then
+                    # return True to say we can't fuse.
+                    if (all([x.name in indices or
+                        x.name not in derived_from_indices for x in ind_seq])):
+                        # Get position in index tuple where parfor indices used.
+                        new_index_positions = [x.name in indices for x in ind_seq]
+                        # Make sure that we aren't accessing a given array with
+                        # different indices in a different order.
+                        return add_check_position(new_index_positions,
+                                                  array_accessed,
+                                                  index_positions,
+                                                  indexed_arrays,
+                                                  non_indexed_arrays)
+                    else:
+                        # index derived from a parfor index used so no fusion
+                        return True
+                else:
+                    # Don't know how the index tuple is built so
+                    # have to assume fusion can't happen.
+                    return True
+            else:
+                # Should be for 1D arrays.
+                if stmt_index.name in indices:
+                    # Array indexed by a parfor index variable.
+                    # Make sure this 1D access is consistent with prior ones.
+                    return add_check_position((True,),
+                                              array_accessed,
+                                              index_positions,
+                                              indexed_arrays,
+                                              non_indexed_arrays)
+                elif stmt_index.name in derived_from_indices:
+                    # If we ever index an array with something calculated
+                    # from an index then no fusion.
+                    return True
+                else:
+                    # Some kind of index that isn't a parfor index or
+                    # one derived from one, e.g., a constant.  Make sure
+                    # this is consistent with prior accessed of this array.
+                    return add_check_position((False,),
+                                              array_accessed,
+                                              index_positions,
+                                              indexed_arrays,
+                                              non_indexed_arrays)
+        else:
+            # We don't know how to handle non-Var indices so no fusion.
+            return True
+
+        # All branches above should cover all the cases and each should
+        # return so we should never get here.
+        assert 0
```

Comment:
Should this raise a `numba.core.errors.InternalError` with a message about this being an illegal state?

Reply:
Done.